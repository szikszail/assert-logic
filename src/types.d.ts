
export type AnyValue = void | boolean | string | object | number | null;
export type AsyncAnyValue = Promise<AnyValue>;
export type AssertionFunction = () => AnyValue | AsyncAnyValue;
export type AssertionValue = AssertionFunction | AnyValue | AsyncAnyValue;
