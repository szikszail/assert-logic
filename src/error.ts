import {Assertion} from "./assertion/assertion";
import {stringify} from "yaml";
import {AnyValue, EvaluationResult} from "./types";
import {valueToString} from "./assertion/pass";

type ErrorRepr = Record<string, string | (ErrorRepr | string)[]>;

function getErrorRepresentation(operator: string, expectation: string, assertion: Assertion, results?: EvaluationResult[]): ErrorRepr {
  const repr: ErrorRepr = {
    [`AssertionError (${operator})`]: expectation,
  };
  if (results?.length > 0) {
    repr.Results = results.map((result) => {
      if (result instanceof AssertionError) {
        return getErrorRepresentation(result.operator, result.expectation, result.assertion, result.results);
      }
      if (result instanceof Error) {
        return {
          // eslint-disable-next-line no-control-regex
          Error: result.message.replace(/\u001b\[\d+m/g, ''),
        };
      }
      return "Pass";
    });
  }
  repr.Expression = assertion.toString();
  return repr;
}

export class AssertionError extends Error {
  constructor(
    public readonly operator: string,
    public readonly expectation: string,
    public readonly assertion: Assertion,
    public readonly results?: EvaluationResult[],
  ) {
    super(stringify(
      getErrorRepresentation(operator, expectation, assertion, results),
      {
        indent: 2,
        // defaultStringType: "PLAIN",
        // defaultKeyType: "PLAIN",
        collectionStyle: 'block',
        // lineWidth: 0,
        // blockQuote: true,
        doubleQuotedMinMultiLineLength: Infinity,
      }
    ));
    this.name = "AssertionError";
  }
}

export class FailedAssertionError extends Error {
  constructor(value: AnyValue) {
    super(`Failed expression: ${valueToString(value)}`);
    this.name = "FailedAssertionError";
  }
}