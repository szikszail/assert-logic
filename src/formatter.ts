import type {AssertionValue} from "./types";

export function toString(value: AssertionValue): string {
  if (value instanceof Function) {
    let name = value?.name;
    if (!name) {
      name = value.toString();
    }
    if (name.length > 20) {
      name = name.slice(0, 17) + "...";
    }
    return `function(${name})`;
  }
  if (value instanceof Promise) {
    return value.toString();
  }
  return `value(${value})`;
}