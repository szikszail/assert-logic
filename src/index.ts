// import debug = require("debug");

// const log = debug("assert-logic");

import type { AssertionValue } from "./types";
import { toString } from "./formatter";

class Assertion {
  evaluate(): void | Promise<void> {
    // pass
  }

  toString(): string {
    return "true";
  }
}

class AssertionError extends Error {
  constructor(statement: string, assertion: Assertion) {
    super(statement + ": " + assertion.toString());
    this.name = "AssertionError";
  }
}

export class AtomicAssertion extends Assertion {
  constructor(private readonly value: AssertionValue) {
    super();
    if (this.value instanceof Assertion) {
      return this.value as AtomicAssertion;
    }
  }

  evaluate(): void | Promise<void> {
    let result = this.value;
    if (result instanceof Function) {
      result = result();
    }
    if (result instanceof Promise) {
      return result.then((value) => {
        if (!value) {
          throw new AssertionError("Assertion failed", this);
        }
      });
    }
    if (!result) {
      throw new AssertionError("Assertion failed", this);
    }
  }

  toString(): string {
    return `${toString(this.value)} to be true`;
  }
}