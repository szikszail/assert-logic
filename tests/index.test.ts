import { AtomicAssertion } from "../src";

describe("assert-logic", () => {
  describe("AtomicAssertion", () => {
    describe("constructor", () => {
      it("should return the value if it is an instance of Assertion", () => {
        const original = new AtomicAssertion(() => true);
        const value = new AtomicAssertion(original);
        expect(value).toBeInstanceOf(AtomicAssertion);
        expect(value).toBe(original);
      });

      it("should return a new instance of AtomicAssertion", () => {
        const value = new AtomicAssertion(() => true);
        expect(value).toBeInstanceOf(AtomicAssertion);
      });

      it("should return a new instance of AtomicAssertion", () => {
        const value = new AtomicAssertion(true);
        expect(value).toBeInstanceOf(AtomicAssertion);
      });
    });

    describe("evaluate", () => {
      it("should return undefined if the value is a function that returns true", () => {
        const value = new AtomicAssertion(() => true);
        const result = value.evaluate();
        expect(result).toBeUndefined();
      });

      it("should return undefined if the value is a function that returns a promise that resolves to true", async () => {
        const value = new AtomicAssertion(() => Promise.resolve(true));
        const result = await value.evaluate();
        expect(result).toBeUndefined();
      });

      it("should throw an AssertionError if the value is a function that returns false", () => {
        const value = new AtomicAssertion(() => false);
        expect(() => value.evaluate()).toThrowErrorMatchingSnapshot();
      });

      it("should throw an AssertionError if the value is a function that returns a promise that resolves to false", async () => {
        const value = new AtomicAssertion(() => Promise.resolve(false));
        await expect(value.evaluate()).rejects.toMatchSnapshot();
      });

      it("should return undefined if the value is true", () => {
        const value = new AtomicAssertion(true);
        const result = value.evaluate();
        expect(result).toBeUndefined();
      });

      it("should throw an AssertionError if the value is false", () => {
        const value = new AtomicAssertion(false);
        expect(() => value.evaluate()).toThrowErrorMatchingSnapshot();
      });

      it("should return undefined if the value is a promise that resolves to true", async () => {
        const value = new AtomicAssertion(Promise.resolve(true));
        const result = await value.evaluate();
        expect(result).toBeUndefined();
      });

      it("should throw an AssertionError if the value is a promise that resolves to false", async () => {
        const value = new AtomicAssertion(Promise.resolve(false));
        await expect(value.evaluate()).rejects.toMatchSnapshot();
      });

      it("should throw an AssertionError if the value is a promise that rejects", async () => {
        const value = new AtomicAssertion(Promise.reject("error"));
        await expect(value.evaluate()).rejects.toMatchSnapshot();
      });

      it("should return undefined if the value is a promise that resolves to a promise that resolves to true", async () => {
        const value = new AtomicAssertion(Promise.resolve(Promise.resolve(true)));
        const result = await value.evaluate();
        expect(result).toBeUndefined();
      });

      // for undefined it should fail
      it("should throw an AssertionError if the value is a promise that resolves to a promise that resolves to false", async () => {
        const value = new AtomicAssertion(Promise.resolve(Promise.resolve(false)));
        await expect(value.evaluate()).rejects.toMatchSnapshot();
      });
    });

    describe("toString", () => {
      it("should return the string representation of an arrow function", () => {
        const value = new AtomicAssertion(() => true);
        const result = value.toString();
        expect(result).toBe("function(() => true) to be true");
      });

      it("should return the string representation of the value if is is a named function", () => {
        const value = new AtomicAssertion(function namedFunction() { return true; });
        const result = value.toString();
        expect(result).toBe("function(namedFunction) to be true");
      });

      it("should return the string representation of the value if it is a async function", () => {
        const value = new AtomicAssertion(async () => true);
        const result = value.toString();
        expect(result).toBe("function(() => tslib_1.__a...) to be true");
      });

      it("should return the string representation of the value if it is a function", () => {
        const value = new AtomicAssertion(true);
        const result = value.toString();
        expect(result).toBe("value(true) to be true");
      });

      it("should return the string representation of the value if it is a function", () => {
        const value = new AtomicAssertion(false);
        const result = value.toString();
        expect(result).toBe("value(false) to be true");
      });

      /*
      it("should return the string representation of the value if it is a function", () => {
        const value = new AtomicAssertion(Promise.resolve(true));
        const result = value.toString();
        expect(result).toBe("true to be true");
      });

      it("should return the string representation of the value if it is a function", () => {
        const value = new AtomicAssertion(Promise.resolve(false));
        const result = value.toString();
        expect(result).toBe("false to be true");
      });

      it("should return the string representation of the value if it is a function", () => {
        const value = new AtomicAssertion(Promise.reject("error"));
        const result = value.toString();
        expect(result).toBe("Promise to be true");
      });

      it("should return the string representation of the value if it is a function", () => {
        const value = new AtomicAssertion(Promise.resolve(Promise.resolve(true)));
        const result = value.toString();
        expect(result).toBe("Promise to be true");
      });

      it("should return the string representation of the value if it is a function", () => {
        const value = new AtomicAssertion(Promise.resolve(Promise.resolve(false)));
        const result = value.toString();
        expect(result).toBe("Promise to be true");
      });
      */
    });
  })
});