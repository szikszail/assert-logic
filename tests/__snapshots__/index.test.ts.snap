// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`assert-logic AtomicAssertion evaluate should throw an AssertionError if the value is a function that returns a promise that resolves to false 1`] = `[AssertionError: Assertion failed: function(() => Promise.res...) to be true]`;

exports[`assert-logic AtomicAssertion evaluate should throw an AssertionError if the value is a function that returns false 1`] = `"Assertion failed: function(() => false) to be true"`;

exports[`assert-logic AtomicAssertion evaluate should throw an AssertionError if the value is a promise that rejects 1`] = `"error"`;

exports[`assert-logic AtomicAssertion evaluate should throw an AssertionError if the value is a promise that resolves to a promise that resolves to false 1`] = `[AssertionError: Assertion failed: [object Promise] to be true]`;

exports[`assert-logic AtomicAssertion evaluate should throw an AssertionError if the value is a promise that resolves to false 1`] = `[AssertionError: Assertion failed: [object Promise] to be true]`;

exports[`assert-logic AtomicAssertion evaluate should throw an AssertionError if the value is false 1`] = `"Assertion failed: value(false) to be true"`;
